src/iop/lut3d.c

```diff
@@ -0,0 +1,1200 @@
+    free(line);
+    fclose(cube_file);
+    return 0;
+  }
+  *clut = lclut;
+//printf("cube ok - level %d, buf_size %d nb colors %d clut %p\n", level, buf_size, i, lclut);
+  free(line);
+  fclose(cube_file);
+  return level;
+}
+
+static inline void lin_to_logFuji(float *lin, float *lout)
+{
+  for (int ch=0; ch<3; ch++)
+  {
+    const float a = 0.555556f * 7.28132488f;
+    const float b = 0.009468f;
+    const float c = 0.344676f;
+    const float d = 0.790453f;
+    const float e = 8.735631f * 7.28132488f;
+    const float f = 0.092864f;
+    const float cut = 0.00089f / 7.28132488f;
+    float logNorm;
+
+    if (lin[ch] >= cut)
+      logNorm = c * log10f(a * lin[ch] +b) + d;
+    else
+      logNorm = e * lin[ch] + f;
+    if( logNorm < 0.0f || lin[ch] <= 0.0f)
+      lout[ch] = 0.0f;
+    else
+      lout[ch] = logNorm;
+  }
+}
+
+static inline void logFuji_to_lin(float *lin, float *lout)
+{
+  for (int ch=0; ch<3; ch++)
+  {
+    const float a = 0.555556f * 7.28132488f;
+    const float b = 0.009468f;
+    const float c = 0.344676f;
+    const float d = 0.790453f;
+    const float e = 8.735631f * 7.28132488f;
+    const float f = 0.092864f;
+    const float cut = 0.100537775223865;
+    float Norm;
+
+    if (lin[ch] >= cut)
+      Norm = (powf(10, (lin[ch] - d) / c) - b) / a;
+    else Norm = (lin[ch] - f) / e;
+    lout[ch] = Norm < 0.0f ? 0.0f : Norm > 1.0f ? 1.0f : Norm;
+  }
+}
+
+// From data/kernels/extended.cl
+static inline float fastlog2(float x)
+{
+  union { float f; unsigned int i; } vx = { x };
+  union { unsigned int i; float f; } mx = { (vx.i & 0x007FFFFF) | 0x3f000000 };
+
+  float y = vx.i;
+
+  y *= 1.1920928955078125e-7f;
+
+  return y - 124.22551499f
+    - 1.498030302f * mx.f
+    - 1.72587999f / (0.3520887068f + mx.f);
+}
+
+static inline void lin_to_log2(float *lin, float *lout, const float middle_grey,
+                                    const float min_exposure, const float dynamic_range)
+{
+  for (int ch=0; ch<3; ch++)
+  {
+    const float logNorm = (fastlog2(lin[ch] / middle_grey) - min_exposure) / dynamic_range;
+    if( logNorm < 0.0f || lin[ch] <= 0.0f)
+    {
+      lout[ch] = 0.0f;
+    }
+    else
+    {
+      lout[ch] = logNorm;
+    }
+  }
+}
+
+static inline void log2_to_lin(float *lin, float *lout, const float middle_grey,
+                                    const float min_exposure, const float dynamic_range)
+{
+  for (int ch=0; ch<3; ch++)
+  {
+    const float Norm = middle_grey * powf(2, lin[ch] * dynamic_range + min_exposure);
+    lout[ch] = Norm < 0.0f ? 0.0f : Norm > 1.0f ? 1.0f : Norm;
+  }
+}
+
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const ibuf, void *const obuf,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+//printf("process\n");
+  dt_iop_lut3d_params_t *p = (dt_iop_lut3d_params_t *)self->params;
+  dt_iop_lut3d_global_data_t *gp = (dt_iop_lut3d_global_data_t *)self->data;

```

Comment:
Instead of self->params should be:

dt_iop_lut3d_data_t *d = (dt_iop_lut3d_data_t *)(piece->data);

and remove the global data.

Reply:
You attach the data to piece instead of module. Just to understand, why that ?
Is that to simplify the code (and to defer calculation from xx_callback() to commit()) ?
Or does that solve an issue ?

Reply:
Not just me, this is how is done. This is to make the pipe thread safe, so the parameters can be updated while the pipe is running.
You're having problems with the pipe right now, let's first make this right and then we can focus on the transforms is still there are issues.

Reply:
> This is to make the pipe thread safe, so the parameters can be updated while the pipe is running.

Great ! Thanks.